File: app/src/main/java/io/shelfy/architecture/domain/usecase/getmovies/GetPopularMoviesUseCase.java
package io.shelfy.architecture.domain.usecase.getmovies;

import java.util.List;

import io.reactivex.rxjava3.core.Single;
import io.shelfy.architecture.domain.entity.Movie;

public interface GetPopularMoviesUseCase {
    Single<List<Movie>> getMovies();
}

File: app/src/main/java/io/shelfy/architecture/domain/usecase/GetPopularMoviesUseCase.kt
package io.shelfy.architecture.domain.usecase

import com.vladk.architecture.domain.MoviesRepository
import com.vladk.architecture.domain.RepositoryCallback
import com.vladk.architecture.domain.model.Movie

interface GetPopularMoviesUseCase {

    fun getMovies(callback: (GetMoviesResult) -> Unit)
}

class GetPopularMoviesUseCaseImpl(
    private val moviesRepository: MoviesRepository
) : GetPopularMoviesUseCase {

    override fun getMovies(callback: (GetMoviesResult) -> Unit) {
        moviesRepository.getPopularMovies(object : RepositoryCallback {
            override fun onSuccess(movies: List<Movie>) {
                val successResult = GetMoviesResult.Success(movies)
                callback(successResult)
            }

            override fun onError(message: String) {
                val errorResult = GetMoviesResult.Error(message)
                callback(errorResult)
            }
        })
    }
}
